x/staking/keeper/liquid_stake.go

```diff
@@ -0,0 +1,395 @@
+
+	return completionTime
+}
+
+// Cancels a pending tokenize share authorization by removing the lock from the queue
+func (k Keeper) CancelTokenizeShareLockExpiration(ctx sdk.Context, address sdk.AccAddress, completionTime time.Time) {
+	authorizations := k.GetPendingTokenizeShareAuthorizations(ctx, completionTime)
+
+	updatedAddresses := []string{}
+	for _, expiringAddress := range authorizations.Addresses {
+		if address.String() != expiringAddress {
+			updatedAddresses = append(updatedAddresses, expiringAddress)
+		}
+	}
+
+	authorizations.Addresses = updatedAddresses
+	k.SetPendingTokenizeShareAuthorizations(ctx, completionTime, authorizations)
+}
+
+// Unlocks all queued tokenize share authorizations that have matured
+// (i.e. have waited the full unbonding period)
+func (k Keeper) RemoveExpiredTokenizeShareLocks(ctx sdk.Context, blockTime time.Time) (unlockedAddresses []string) {
+	store := ctx.KVStore(k.storeKey)
+
+	// iterators all time slices from time 0 until the current block time
+	prefixEnd := sdk.InclusiveEndBytes(types.GetTokenizeShareAuthorizationTimeKey(blockTime))
+	iterator := store.Iterator(types.TokenizeSharesUnlockQueuePrefix, prefixEnd)
+	defer iterator.Close()
+
+	unlockedAddresses = []string{}
+	for ; iterator.Valid(); iterator.Next() {
+		authorizations := types.PendingTokenizeShareAuthorizations{}
+		k.cdc.MustUnmarshal(iterator.Value(), &authorizations)
+
+		for _, addressString := range authorizations.Addresses {
+			k.RemoveTokenizeSharesLock(ctx, sdk.MustAccAddressFromBech32(addressString))
+			unlockedAddresses = append(unlockedAddresses, addressString)
+		}
+		store.Delete(iterator.Key())
+	}
+
+	return unlockedAddresses
+}
+
+// Calculates and sets the global liquid staked tokens and liquid shares by validator
+// The totals are determined by looping each delegation record and summing the stake
+// if the delegator has a 32-length address. Checking for a 32-length address will capture
+// ICA accounts, as well as tokenized delegations which are owned by module accounts
+// under the hood
+// This function must be called in the upgrade handler which onboards LSM, as
+// well as any time the liquid staking cap is re-enabled
+func (k Keeper) RefreshTotalLiquidStaked(ctx sdk.Context) error {
+	// First reset each validator's liquid shares to 0
+	for _, validator := range k.GetAllValidators(ctx) {
+		validator.LiquidShares = sdk.ZeroDec()
+		k.SetValidator(ctx, validator)
+	}
+
+	// Sum up the total liquid tokens and increment each validator's liquid shares
+	totalLiquidStakedTokens := sdk.ZeroInt()
+	for _, delegation := range k.GetAllDelegations(ctx) {
+		delegatorAddress, err := sdk.AccAddressFromBech32(delegation.DelegatorAddress)
+		if err != nil {
+			return err
+		}
+		validatorAddress, err := sdk.ValAddressFromBech32(delegation.ValidatorAddress)
+		if err != nil {
+			return err
+		}
+
+		validator, found := k.GetValidator(ctx, validatorAddress)
```

Comment:
nit: This can be moved inside the if branch bellow.

Reply:
[done](https://github.com/cosmos/cosmos-sdk/pull/16747/commits/73e72b9ef30f2315adeeefa33aa2526cf23ba8b7)